Src/Newtonsoft.Json/JsonSerializer.cs

```diff
@@ -1009,9 +1009,10 @@ private void ResetReader(JsonReader reader, CultureInfo previousCulture, DateTim
                 reader.DateFormatString = previousDateFormatString;
             }
 
-            if (reader is JsonTextReader textReader)
+            if (reader is JsonTextReader textReader && textReader.PropertyNameTable != null &&
+                _contractResolver is DefaultContractResolver resolver && textReader.PropertyNameTable == resolver.GetNameTable())
             {
-                textReader.NameTable = null;
+                textReader.PropertyNameTable = null;
```

Comment:
Add two unit tests:
1. `PropertyNameTable` is set back to null when the contract resolver `NameTable` is used and serialization is complete.
2. `PropertyNameTable` is not set back to null when a custom `NameTable` is used.

Reply:
I think you already have 2.

Reply:
I already have both tests covered. See L209 and L234 in JsonConverterTest.cs